api/envoy/service/discovery/v3/discovery.proto

```diff
@@ -43,6 +43,14 @@ message DiscoveryRequest {
   // which will be explicitly enumerated in resource_names.
   repeated string resource_names = 3;
 
+  // The resource namespace for this subscription. When *resource_names* is
+  // empty, this will be used by the management server to determine the set of
+  // resources to deliver. If *resource_names* is non-empty, all named resources
+  // must be prefixed with *namespace*. See the xDS protocol specification for
```

Comment:
I think you mean here that the management server should prefix what is found in the namespace field with each resource name, vs. the client doing it and having that be the final name in resource_names? If the former I agree, but can you clarify slightly as it's a little ambiguous.

Reply:
The prefix rule is an odd constraint here if the namespace is also the name for a collection of resource_names as indicated by 'determine the set of resources to deliver'.   

If you want to model 'give me the set of listeners for proxy X' I wouldn't call X a namespace. It sounds more like a 'Collection' in classical RESTful terminology. The prefixing requirement is odd because the name of one of the listeners could be defined by a different entity than the one that defined X.  This gets worse if you want a Collection to refer to resources that span authorities.

Reply:
The authority example is an interesting one.  What if we said that the authority of the resource and the authority of the namespace can be decoupled?  In other words, the authority of the namespace gets to say which resources are included in that namespace, but the individual resources can be controlled by different authorities.  In effect, the namespace would basically be like a Unix directory, and the entries in it would basically be like hard links: the permissions on the directory control what entries are in the directory, but the permissions on the individual files control access to the individual files, and the two can be completely different.

It seems relevant here to mention that @htuch and I have discussed the idea of using attributes instead of namespaces, where every resource can have a set of attributes (key/value pairs), and the client can request all resources that match a particular set of attributes.  The main advantage of that approach over namespaces is that it would more easily allow the same resource to be requested by different criteria.  However, neither of us could come up with a concrete use-case where that property would be useful, so our consensus was to stick with namespaces until and unless we could come up with such a use-case.  But it's worth noting that the "hard link" approach I described above would also address this use-case if we ever run into it.

Reply:
I'm not a fan of the hard-link analogy as it implies the existence of a name mapping scheme in one authority that is aware of the naming scheme in another. I prefer the idea of composing at point-of-use with fully qualified names. 

The outer-resource can decide whether to delegate or not so behavior is constrained without having to overlay access control. Leaf level access is constrained by the normal rules of the authority

My analogy is a web page, you served the following from
xds://some.authority/listeners/for/some/proxy   (pardon the shorthand)

<listener-list>
  <listener>{ "some" : "config"}</listener>
  <listener src="xds://some.other.authority/listeners/some/listener/resource"/>
</listener-list>

and Envoy understand this and how to resolve delegated resource references across authority boundaries. 

 I think @htuch is going to write something up along these lines.

Reply:
Also composition at point-of-use aligns better with credentialing. 

Assuming different authorities require different credentials to read their resources. You wouldn't want the presentation of a credential valid for one to imply that it should be able to read a resource in another.

Control plane implementations are free to replicate resources from other control planes etc but I think that should be entirely opaque to Envoy and doesn't require any support in these APIs.